[Transform] Regression: continuous transform fails after start with [composite] failed to parse field [sources]

Creating a continuous transform with 1 histogram or date histogram succeeds but fails after `_start`. This is a regression introduced by #55287.

Repro:

```
PUT _transform/histogram-pivot
{
  "source": {
    "index": "kibana_sample_data_logs"
  },
  "dest": {
    "index": "kibana_sample_data_logs-pivot"
  },
  "sync": {
    "time": {
      "field": "timestamp"
    }
  }, 
  "pivot": {
    "group_by": {
      "h": {
        "histogram": {
          "field": "bytes",
          "interval": 50
          }
      }
    },
    "aggregations": {
      "max": {
        "max": {
          "field": "bytes"
        }
      }
    }
  }
}
```

failure:

```
org.elasticsearch.common.xcontent.XContentParseException: [1:13] [composite] failed to parse field [sources]
        at org.elasticsearch.common.xcontent.ObjectParser.parseValue(ObjectParser.java:531) ~[elasticsearch-x-content-7.9.0-SNAPSHOT.jar:7.9.0-SNAPSHOT]
        at org.elasticsearch.common.xcontent.ObjectParser.parseArray(ObjectParser.java:523) ~[elasticsearch-x-content-7.9.0-SNAPSHOT.jar:7.9.0-SNAPSHOT]
        at org.elasticsearch.common.xcontent.ObjectParser.parseSub(ObjectParser.java:555) ~[elasticsearch-x-content-7.9.0-SNAPSHOT.jar:7.9.0-SNAPSHOT]
        at org.elasticsearch.common.xcontent.ObjectParser.parse(ObjectParser.java:324) ~[elasticsearch-x-content-7.9.0-SNAPSHOT.jar:7.9.0-SNAPSHOT]
        at org.elasticsearch.common.xcontent.ConstructingObjectParser.parse(ConstructingObjectParser.java:171) ~[elasticsearch-x-content-7.9.0-SNAPSHOT.jar:7.9.0-SNAPSHOT]
        at org.elasticsearch.xpack.transform.transforms.pivot.Pivot.createCompositeAggregationSources(Pivot.java:361) ~[?:?]
        at org.elasticsearch.xpack.transform.transforms.pivot.Pivot.buildChangeCollector(Pivot.java:219) ~[?:?]
        at org.elasticsearch.xpack.transform.transforms.TransformIndexer.initializeFunction(TransformIndexer.java:349) ~[?:?]
        at org.elasticsearch.xpack.transform.transforms.TransformIndexer.lambda$onStart$10(TransformIndexer.java:249) ~[?:?]
        at org.elasticsearch.action.ActionListener$1.onResponse(ActionListener.java:63) [elasticsearch-7.9.0-SNAPSHOT.jar:7.9.0-SNAPSHOT]
        at org.elasticsearch.xpack.transform.transforms.TransformIndexer.lambda$onStart$11(TransformIndexer.java:300) [transform-7.9.0-SNAPSHOT.jar:7.9.0-SNAPSHOT]
        at org.elasticsearch.action.ActionListener$1.onResponse(ActionListener.java:63) [elasticsearch-7.9.0-SNAPSHOT.jar:7.9.0-SNAPSHOT]
        at org.elasticsearch.xpack.transform.persistence.IndexBasedTransformConfigManager.parseTransformLenientlyFromSource(IndexBasedTransformConfigManager.java:666) [transform-7.9.0-SNAPSHOT.jar:7.9.0-SNAPSHOT]
        at org.elasticsearch.xpack.transform.persistence.IndexBasedTransformConfigManager.lambda$getTransformConfiguration$7(IndexBasedTransformConfigManager.java:355) [transform-7.9.0-SNAPSHOT.jar:7.9.0-SNAPSHOT]
        at org.elasticsearch.action.ActionListener$1.onResponse(ActionListener.java:63) [elasticsearch-7.9.0-SNAPSHOT.jar:7.9.0-SNAPSHOT]
        at org.elasticsearch.action.support.ContextPreservingActionListener.onResponse(ContextPreservingActionListener.java:43) [elasticsearch-7.9.0-SNAPSHOT.jar:7.9.0-SNAPSHOT]
        at org.elasticsearch.action.support.TransportAction$1.onResponse(TransportAction.java:89) [elasticsearch-7.9.0-SNAPSHOT.jar:7.9.0-SNAPSHOT]
        at org.elasticsearch.action.support.TransportAction$1.onResponse(TransportAction.java:83) [elasticsearch-7.9.0-SNAPSHOT.jar:7.9.0-SNAPSHOT]
        at org.elasticsearch.action.search.AbstractSearchAsyncAction.sendSearchResponse(AbstractSearchAsyncAction.java:545) [elasticsearch-7.9.0-SNAPSHOT.jar:7.9.0-SNAPSHOT]
        at org.elasticsearch.action.search.ExpandSearchPhase.run(ExpandSearchPhase.java:117) [elasticsearch-7.9.0-SNAPSHOT.jar:7.9.0-SNAPSHOT]
        at org.elasticsearch.action.search.AbstractSearchAsyncAction.executePhase(AbstractSearchAsyncAction.java:350) [elasticsearch-7.9.0-SNAPSHOT.jar:7.9.0-SNAPSHOT]
        at org.elasticsearch.action.search.AbstractSearchAsyncAction.executeNextPhase(AbstractSearchAsyncAction.java:344) [elasticsearch-7.9.0-SNAPSHOT.jar:7.9.0-SNAPSHOT]
        at org.elasticsearch.action.search.FetchSearchPhase.moveToNextPhase(FetchSearchPhase.java:231) [elasticsearch-7.9.0-SNAPSHOT.jar:7.9.0-SNAPSHOT]
        at org.elasticsearch.action.search.FetchSearchPhase.lambda$innerRun$1(FetchSearchPhase.java:119) [elasticsearch-7.9.0-SNAPSHOT.jar:7.9.0-SNAPSHOT]
        at org.elasticsearch.action.search.FetchSearchPhase.innerRun(FetchSearchPhase.java:125) [elasticsearch-7.9.0-SNAPSHOT.jar:7.9.0-SNAPSHOT]
        at org.elasticsearch.action.search.FetchSearchPhase.access$000(FetchSearchPhase.java:47) [elasticsearch-7.9.0-SNAPSHOT.jar:7.9.0-SNAPSHOT]
        at org.elasticsearch.action.search.FetchSearchPhase$1.doRun(FetchSearchPhase.java:95) [elasticsearch-7.9.0-SNAPSHOT.jar:7.9.0-SNAPSHOT]
        at org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:37) [elasticsearch-7.9.0-SNAPSHOT.jar:7.9.0-SNAPSHOT]
        at org.elasticsearch.common.util.concurrent.TimedRunnable.doRun(TimedRunnable.java:44) [elasticsearch-7.9.0-SNAPSHOT.jar:7.9.0-SNAPSHOT]
        at org.elasticsearch.common.util.concurrent.ThreadContext$ContextPreservingAbstractRunnable.doRun(ThreadContext.java:710) [elasticsearch-7.9.0-SNAPSHOT.jar:7.9.0-SNAPSHOT]
        at org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:37) [elasticsearch-7.9.0-SNAPSHOT.jar:7.9.0-SNAPSHOT]
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1130) [?:?]
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:630) [?:?]
        at java.lang.Thread.run(Thread.java:832) [?:?]
Caused by: org.elasticsearch.common.xcontent.XContentParseException: Failed to build [composite] after last required field arrived
        at org.elasticsearch.common.xcontent.ConstructingObjectParser$Target.buildTarget(ConstructingObjectParser.java:534) ~[elasticsearch-x-content-7.9.0-SNAPSHOT.jar:7.9.0-SNAPSHOT]
        at org.elasticsearch.common.xcontent.ConstructingObjectParser$Target.constructorArg(ConstructingObjectParser.java:460) ~[elasticsearch-x-content-7.9.0-SNAPSHOT.jar:7.9.0-SNAPSHOT]
        at org.elasticsearch.common.xcontent.ConstructingObjectParser$Target.access$900(ConstructingObjectParser.java:408) ~[elasticsearch-x-content-7.9.0-SNAPSHOT.jar:7.9.0-SNAPSHOT]
        at org.elasticsearch.common.xcontent.ConstructingObjectParser.lambda$declareField$3(ConstructingObjectParser.java:220) ~[elasticsearch-x-content-7.9.0-SNAPSHOT.jar:7.9.0-SNAPSHOT]
        at org.elasticsearch.common.xcontent.ObjectParser.lambda$declareField$9(ObjectParser.java:386) ~[elasticsearch-x-content-7.9.0-SNAPSHOT.jar:7.9.0-SNAPSHOT]
        at org.elasticsearch.common.xcontent.ObjectParser.parseValue(ObjectParser.java:529) ~[elasticsearch-x-content-7.9.0-SNAPSHOT.jar:7.9.0-SNAPSHOT]
        ... 33 more
Caused by: java.lang.IllegalArgumentException: Composite [sources] cannot be null or empty
        at org.elasticsearch.search.aggregations.bucket.composite.CompositeAggregationBuilder.validateSources(CompositeAggregationBuilder.java:177) ~[elasticsearch-7.9.0-SNAPSHOT.jar:7.9.0-SNAPSHOT]
        at org.elasticsearch.search.aggregations.bucket.composite.CompositeAggregationBuilder.<init>(CompositeAggregationBuilder.java:70) ~[elasticsearch-7.9.0-SNAPSHOT.jar:7.9.0-SNAPSHOT]
        at org.elasticsearch.search.aggregations.bucket.composite.CompositeAggregationBuilder.lambda$static$0(CompositeAggregationBuilder.java:55) ~[elasticsearch-7.9.0-SNAPSHOT.jar:7.9.0-SNAPSHOT]
        at org.elasticsearch.common.xcontent.ConstructingObjectParser$Target.buildTarget(ConstructingObjectParser.java:521) ~[elasticsearch-x-content-7.9.0-SNAPSHOT.jar:7.9.0-SNAPSHOT]
        at org.elasticsearch.common.xcontent.ConstructingObjectParser$Target.constructorArg(ConstructingObjectParser.java:460) ~[elasticsearch-x-content-7.9.0-SNAPSHOT.jar:7.9.0-SNAPSHOT]
        at org.elasticsearch.common.xcontent.ConstructingObjectParser$Target.access$900(ConstructingObjectParser.java:408) ~[elasticsearch-x-content-7.9.0-SNAPSHOT.jar:7.9.0-SNAPSHOT]
        at org.elasticsearch.common.xcontent.ConstructingObjectParser.lambda$declareField$3(ConstructingObjectParser.java:220) ~[elasticsearch-x-content-7.9.0-SNAPSHOT.jar:7.9.0-SNAPSHOT]
        at org.elasticsearch.common.xcontent.ObjectParser.lambda$declareField$9(ObjectParser.java:386) ~[elasticsearch-x-content-7.9.0-SNAPSHOT.jar:7.9.0-SNAPSHOT]
        at org.elasticsearch.common.xcontent.ObjectParser.parseValue(ObjectParser.java:529) ~[elasticsearch-x-content-7.9.0-SNAPSHOT.jar:7.9.0-SNAPSHOT]
        ... 33 more

```

--- a/x-pack/plugin/transform/qa/single-node-tests/src/test/java/org/elasticsearch/xpack/transform/integration/TransformPivotRestIT.java
+++ b/x-pack/plugin/transform/qa/single-node-tests/src/test/java/org/elasticsearch/xpack/transform/integration/TransformPivotRestIT.java
@@ -500,6 +500,115 @@ public class TransformPivotRestIT extends TransformRestTestCase {
         assertOnePivotValue(transformIndex + "/_search?q=every_2:0.0", 1.0);
     }
 
+    public void testContinuousPivotHistogram() throws Exception {
+        String indexName = "continuous_reviews_histogram";
+        createReviewsIndex(indexName);
+        String transformId = "simple_continuous_pivot";
+        String transformIndex = "pivot_reviews_continuous_histogram";
+        setupDataAccessRole(DATA_ACCESS_ROLE, indexName, transformIndex);
+        final Request createTransformRequest = createRequestWithAuth(
+            "PUT",
+            getTransformEndpoint() + transformId,
+            BASIC_AUTH_VALUE_TRANSFORM_ADMIN_WITH_SOME_DATA_ACCESS
+        );
+        String config = "{"
+            + " \"source\": {\"index\":\""
+            + indexName
+            + "\"},"
+            + " \"dest\": {\"index\":\""
+            + transformIndex
+            + "\"},"
+            + " \"frequency\": \"1s\","
+            + " \"sync\": {\"time\": {\"field\": \"timestamp\", \"delay\": \"1s\"}},"
+            + " \"pivot\": {"
+            + "   \"group_by\": {"
+            + "     \"every_2\": {"
+            + "       \"histogram\": {"
+            + "         \"interval\": 2,\"field\":\"stars\""
+            + " } } },"
+            + "   \"aggregations\": {"
+            + "     \"user_dc\": {"
+            + "       \"cardinality\": {"
+            + "         \"field\": \"user_id\""
+            + " } } } }"
+            + "}";
+        createTransformRequest.setJsonEntity(config);
+        Map<String, Object> createTransformResponse = entityAsMap(client().performRequest(createTransformRequest));
+        assertThat(createTransformResponse.get("acknowledged"), equalTo(Boolean.TRUE));
+
+        startAndWaitForContinuousTransform(transformId, transformIndex, null);
+        assertTrue(indexExists(transformIndex));
+
+        // we expect 3 documents as there shall be 5 unique star values and we are bucketing every 2 starting at 0
+        Map<String, Object> indexStats = getAsMap(transformIndex + "/_stats");
+        assertEquals(3, XContentMapValues.extractValue("_all.total.docs.count", indexStats));
+
+        Map<String, Object> searchResult = getAsMap(transformIndex + "/_search?q=every_2:0.0");
+        assertEquals(1, XContentMapValues.extractValue("hits.total.value", searchResult));
+        assertThat(((List<?>) XContentMapValues.extractValue("hits.hits._source.user_dc", searchResult)).get(0), equalTo(19.0));
+
+        searchResult = getAsMap(transformIndex + "/_search?q=every_2:2.0");
+        assertEquals(1, XContentMapValues.extractValue("hits.total.value", searchResult));
+        assertThat(((List<?>) XContentMapValues.extractValue("hits.hits._source.user_dc", searchResult)).get(0), equalTo(27.0));
+
+        searchResult = getAsMap(transformIndex + "/_search?q=every_2:4.0");
+        assertEquals(1, XContentMapValues.extractValue("hits.total.value", searchResult));
+        assertThat(((List<?>) XContentMapValues.extractValue("hits.hits._source.user_dc", searchResult)).get(0), equalTo(27.0));
+
+        final StringBuilder bulk = new StringBuilder();
+
+        long dateStamp = Instant.now().toEpochMilli() - 1_000;
+
+        // add 5 data points with 3 new users: 27, 28, 29
+        for (int i = 25; i < 30; i++) {
+            bulk.append("{\"index\":{\"_index\":\"" + indexName + "\"}}\n");
+            String location = (i + 10) + "," + (i + 15);
+
+            bulk.append("{\"user_id\":\"")
+                .append("user_")
+                .append(i)
+                .append("\",\"business_id\":\"")
+                .append("business_")
+                .append(i)
+                .append("\",\"stars\":")
+                .append(3)
+                .append(",\"location\":\"")
+                .append(location)
+                .append("\",\"timestamp\":")
+                .append(dateStamp)
+                .append("}\n");
+        }
+        bulk.append("\r\n");
+
+        final Request bulkRequest = new Request("POST", "/_bulk");
+        bulkRequest.addParameter("refresh", "true");
+        bulkRequest.setJsonEntity(bulk.toString());
+        client().performRequest(bulkRequest);
+
+        waitForTransformCheckpoint(transformId, 2);
+
+        stopTransform(transformId, false);
+        refreshIndex(transformIndex);
+
+        // assert after changes
+        // still 3 documents
+        indexStats = getAsMap(transformIndex + "/_stats");
+        assertEquals(3, XContentMapValues.extractValue("_all.total.docs.count", indexStats));
+
+        searchResult = getAsMap(transformIndex + "/_search?q=every_2:0.0");
+        assertEquals(1, XContentMapValues.extractValue("hits.total.value", searchResult));
+        assertThat(((List<?>) XContentMapValues.extractValue("hits.hits._source.user_dc", searchResult)).get(0), equalTo(19.0));
+
+        searchResult = getAsMap(transformIndex + "/_search?q=every_2:2.0");
+        assertEquals(1, XContentMapValues.extractValue("hits.total.value", searchResult));
+        assertThat(((List<?>) XContentMapValues.extractValue("hits.hits._source.user_dc", searchResult)).get(0), equalTo(30.0));
+
+        searchResult = getAsMap(transformIndex + "/_search?q=every_2:4.0");
+        assertEquals(1, XContentMapValues.extractValue("hits.total.value", searchResult));
+        assertThat(((List<?>) XContentMapValues.extractValue("hits.hits._source.user_dc", searchResult)).get(0), equalTo(27.0));
+
+    }
+
     public void testBiggerPivot() throws Exception {
         String transformId = "bigger_pivot";
         String transformIndex = "bigger_pivot_reviews";

--- a/x-pack/plugin/transform/src/main/java/org/elasticsearch/xpack/transform/transforms/TransformIndexer.java
+++ b/x-pack/plugin/transform/src/main/java/org/elasticsearch/xpack/transform/transforms/TransformIndexer.java
@@ -345,7 +345,7 @@ public abstract class TransformIndexer extends AsyncTwoPhaseIndexer<TransformInd
         // create the function
         function = FunctionFactory.create(getConfig());
 
-        if (isContinuous()) {
+        if (isContinuous() && function.supportsIncrementalBucketUpdate()) {
             changeCollector = function.buildChangeCollector(getConfig().getSyncConfig().getField());
         }
     }
@@ -370,7 +370,7 @@ public abstract class TransformIndexer extends AsyncTwoPhaseIndexer<TransformInd
             // reset the page size, so we do not memorize a low page size forever
             pageSize = function.getInitialPageSize();
             // reset the changed bucket to free memory
-            if (isContinuous()) {
+            if (changeCollector != null) {
                 changeCollector.clear();
             }
 
@@ -749,7 +749,9 @@ public abstract class TransformIndexer extends AsyncTwoPhaseIndexer<TransformInd
         BoolQueryBuilder filteredQuery = new BoolQueryBuilder().filter(queryBuilder)
             .filter(config.getSyncConfig().getRangeQuery(nextCheckpoint));
 
-        QueryBuilder filter = changeCollector.buildFilterQuery(lastCheckpoint.getTimeUpperBound(), nextCheckpoint.getTimeUpperBound());
+        QueryBuilder filter = changeCollector != null
+            ? changeCollector.buildFilterQuery(lastCheckpoint.getTimeUpperBound(), nextCheckpoint.getTimeUpperBound())
+            : null;
 
         if (filter != null) {
             filteredQuery.filter(filter);

